conda_build/jinja_context.py

```diff
@@ -203,14 +205,154 @@ def load_file_regex(config, load_file, regex_pattern, from_recipe_dir=False,
     return match if match else None
 
 
+def pin_compatible(config, package_name, upper_bound=None, pins=('p', ),
+                   permit_undefined_jinja=True):
+    """dynamically pin based on currently installed version.
+
+    only mandatory input is package_name.
+    upper_bound is the authoritative upper bound, if provided.  The lower bound is the the
+        currently installed version.
+    pin expressions can be any string separated by `.`
+        The effective pin is length of the string split on '.'.
+    """
+    compatibility = None
+
+    # this is the version split up into its component bits.
+    # There are two cases considered here (so far):
+    # 1. Good packages that follow semver style (if not philosophy).  For example, 1.2.3
+    # 2. Evil packages that cram everything alongside a single major version.  For example, 9b
+    version = get_installed_packages(config.build_prefix).get(package_name, {}).get('version')
+    if version:
+        if upper_bound:
+            compatibility = ">=" + version + ","
+            compatibility += '<{upper_bound}'.format(upper_bound=upper_bound)
+        else:
+            pins = ensure_list(pins)
+            compatibility = apply_pin_expressions(version, pins)
+
+    if not compatibility and not permit_undefined_jinja:
+        raise RuntimeError("Could not get compatibility information for {} package.  Is the "
+                            "build environment created?".format(package_name))
+    return compatibility
+
+
+# map python version to default compiler on windows, to match upstream python
+#    This mapping only sets the "native" compiler, and can be overridden by specifying a compiler
+#    in the conda-build variant configuration
+compilers = {
+    'win': {
+        'c': {
+            '2.7': 'vs2008',
+            '3.3': 'vs2010',
+            '3.4': 'vs2010',
+            '3.5': 'vs2015',
+        },
+        'cxx': {
+            '2.7': 'vs2008',
+            '3.3': 'vs2010',
+            '3.4': 'vs2010',
+            '3.5': 'vs2015',
+        },
+        'fortran': 'gfortran',
+    },
+    'linux': {
+        'c': 'gcc',
+        'cxx': 'g++',
+        'fortran': 'gfortran',
+    },
+    # TODO: Clang?  System clang, or compiled package?  Can handle either as package.
+    'osx': {
+        'c': 'gcc',
+        'cxx': 'g++',
```

Comment:
IMHO this should be the system toolchain, which would be `clang` and `clang++`.

Apple has been moving towards the elimination of `gcc` for some time. 10.6 was the last version that actually had these as the standalone default system toolchain. By 10.7 they were merely frontends to llvm. By 10.9, `gcc` merely redirects to `clang`, but links with `stdlibc++` for C++ code. It should also be noted that all of the libraries and includes are from 4.2.1 as it was the last version under GPL2 (4.2.2 was under GPL3). So if one uses this as the system compiler, they inherit all the bugs from working with this old version.

ref: http://stackoverflow.com/a/8707446
ref: http://stackoverflow.com/q/19535422
ref: http://unix.stackexchange.com/a/49970

Note: Am not saying people can't install something else (e.g. the `gcc` package) and use that. Just that the default case should assume the user has the system toolchain available and make no other assumptions.